test/jdk/java/util/zip/ZipFile/ZipSourceCache.java

```diff
@@ -0,0 +1,160 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.AfterAll;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ZipSourceCache {
+
+    private static final String ZIPFILE_NAME =
+            System.currentTimeMillis() + "-bug8317678.zip";
+    private static final String ZIPENTRY_NAME = "random.txt";
+    private static final boolean DEBUG = false;
+
+    private static File relativeFile = new File(ZIPFILE_NAME);
+    private static File absoluteFile = new File(ZIPFILE_NAME).getAbsoluteFile();
+    private static boolean hasfileKeySupport;
+
+    @BeforeAll
+    public static void setup() throws Exception {
+        createZipFile("test1");
+        var attrs = Files.readAttributes(relativeFile.toPath(), BasicFileAttributes.class);
+        hasfileKeySupport = (attrs.fileKey() != null);
+    }
+
+    @AfterAll
+    public static void cleanup() throws IOException {
+        Files.deleteIfExists(Path.of(ZIPFILE_NAME));
+    }
+
+    /*
+     * Monitor the internal "files" HashMap to ensure that we only
+     * create one <Key, Source> mapping per unique zip file.
+     *
+     * This test also ensures that a new <Key, Source> mapping is created
+     * if an update to an existing zip file is detected.
+     */
+    @Test
+    public void test() throws Exception {
+        ZipFile absoluteZipFile;
+        HashMap internalMap;
+        int numSources;
+        try (ZipFile zipFile = new ZipFile(ZIPFILE_NAME)) {
+            Class source = Class.forName("java.util.zip.ZipFile$Source");
+            Field filesMap = source.getDeclaredField("files");
+            filesMap.setAccessible(true);
+            internalMap = (HashMap) filesMap.get(zipFile);
+            numSources = internalMap.size();
+            // opening of same zip file shouldn't cause new Source
+            // to be constructed on filesystems which support fileKey()
+            absoluteZipFile = new ZipFile(absoluteFile);
+            if (hasfileKeySupport) {
+                assertEquals(numSources, internalMap.size());
+            } else {
+                assertEquals(++numSources, internalMap.size());
+            }
+
+            // update the zip file, should expect a new Source Object
+            // ignore this part of test if file can't be updated (can't overwrite)
+            if (createZipFile("differentContent")) {
+                ZipFile z = new ZipFile(ZIPFILE_NAME);
+                // update of file should trigger new <Key, Source> mapping
+                assertEquals(++numSources, internalMap.size());
+                // new Source created, CEN structure should map fine
+                readZipFileContents(z);
+                // the old Source in use for old file, should no longer map correctly
+                IOException ioe = assertThrows(IOException.class, () -> readZipFileContents(absoluteZipFile));
+                assertEquals("ZipFile invalid LOC header (bad signature)", ioe.getMessage());
```

Comment:
Perhaps put the expected message in a constant in the event we need to change it at some point it is easier to find